Source/WebCore/css/CSSPrimitiveValue.cpp

```diff
@@ -271,252 +153,73 @@ CSSUnitType CSSPrimitiveValue::primitiveType() const
-    case CSSUnitType::CSS_IC:
-    case CSSUnitType::CSS_PX:
-    case CSSUnitType::CSS_CM:
-    case CSSUnitType::CSS_MM:
-    case CSSUnitType::CSS_IN:
-    case CSSUnitType::CSS_PT:
-    case CSSUnitType::CSS_PC:
-    case CSSUnitType::CSS_DEG:
-    case CSSUnitType::CSS_RAD:
-    case CSSUnitType::CSS_GRAD:
-    case CSSUnitType::CSS_MS:
-    case CSSUnitType::CSS_S:
-    case CSSUnitType::CSS_HZ:
-    case CSSUnitType::CSS_KHZ:
-    case CSSUnitType::CSS_TURN:
-    case CSSUnitType::CSS_VW:
-    case CSSUnitType::CSS_VH:
-    case CSSUnitType::CSS_VMIN:
-    case CSSUnitType::CSS_VMAX:
-    case CSSUnitType::CSS_VB:
-    case CSSUnitType::CSS_VI:
-    case CSSUnitType::CSS_SVW:
-    case CSSUnitType::CSS_SVH:
-    case CSSUnitType::CSS_SVMIN:
-    case CSSUnitType::CSS_SVMAX:
-    case CSSUnitType::CSS_SVB:
-    case CSSUnitType::CSS_SVI:
-    case CSSUnitType::CSS_LVW:
-    case CSSUnitType::CSS_LVH:
-    case CSSUnitType::CSS_LVMIN:
-    case CSSUnitType::CSS_LVMAX:
-    case CSSUnitType::CSS_LVB:
-    case CSSUnitType::CSS_LVI:
-    case CSSUnitType::CSS_DVW:
-    case CSSUnitType::CSS_DVH:
-    case CSSUnitType::CSS_DVMIN:
-    case CSSUnitType::CSS_DVMAX:
-    case CSSUnitType::CSS_DVB:
-    case CSSUnitType::CSS_DVI:
-    case CSSUnitType::CSS_DPPX:
-    case CSSUnitType::CSS_X:
-    case CSSUnitType::CSS_DPI:
-    case CSSUnitType::CSS_DPCM:
-    case CSSUnitType::CSS_FR:
-    case CSSUnitType::CSS_Q:
-    case CSSUnitType::CSS_LHS:
-    case CSSUnitType::CSS_RLHS:
-    case CSSUnitType::CSS_IDENT:
-    case CSSUnitType::CSS_UNKNOWN:
-    case CSSUnitType::CSS_PROPERTY_ID:
-    case CSSUnitType::CSS_VALUE_ID:
-    case CSSUnitType::CSS_CQW:
-    case CSSUnitType::CSS_CQH:
-    case CSSUnitType::CSS_CQI:
-    case CSSUnitType::CSS_CQB:
-    case CSSUnitType::CSS_CQMIN:
-    case CSSUnitType::CSS_CQMAX:
-        ASSERT(!isStringType(type));
-        break;
-    }
     if (m_hasCachedCSSText) {
         ASSERT(serializedPrimitiveValues().contains(this));
         serializedPrimitiveValues().remove(this);
     }
 }
 
-Ref<CSSPrimitiveValue> CSSPrimitiveValue::create(CSSPropertyID propertyID)
-{
-    return adoptRef(*new CSSPrimitiveValue(propertyID));
-}
-
-static CSSPrimitiveValue* valueFromPool(Span<LazyNeverDestroyed<CSSPrimitiveValue>> pool, double value)
+inline CSSPrimitiveValue* CSSPrimitiveValue::scalarValue(double value, CSSUnitType unit)
 {
-    // Casting to a signed integer first since casting a negative floating point value to an unsigned
-    // integer is undefined behavior.
-    unsigned poolIndex = static_cast<unsigned>(static_cast<int>(value));
-    double roundTripValue = poolIndex;
-    if (!memcmp(&value, &roundTripValue, sizeof(double)) && poolIndex < pool.size())
-        return &pool[poolIndex].get();
-    return nullptr;
+    constexpr uintptr_t maxInteger = std::numeric_limits<uintptr_t>::max() >> IntegerShift;
+    uintptr_t integer = static_cast<intptr_t>(value);
+    double roundTripValue = integer;
```

Comment:
Could we use some more generic scheme for compressing doubles into fewer bits? Just allowing integers seems limiting.

Maybe we could even fit all doubles (in 64bit systems) with some acceptable loss of precision/range and avoid having a heap allocated form of CSSPrimitiveValue altogether? That would then allow empty ref/deref when the type is known.

Reply:
I have done some experiments with this and I have something that covers tons of the actual CSS numeric value with these bits. I will do a patch switching to that scheme after landing this one.